File: probs/10942.py
import sys
input = sys.stdin.readline

input = open('input.txt', 'r').readline

N = int(input())
A = [0]+list(map(int, input().split()))

# dp[i][j] -> i번째부터 j번째까지 팰린드롬인지?
dp = [[0]*(N+1) for _ in range(N+1)]
for i in range(1, N+1):
    dp[i][i] = 1

for i in range(1, N+1):
    for j in range(1, i):
        if A[i] == A[j]:
            if i == j+1:
                dp[j][i] = 1
            else:
                dp[j][i] = dp[j+1][i-1]

for _ in range(int(input())):
    S, E = map(int, input().split())
    print(dp[S][E])

File: probs/10942.cpp
#include <bits/stdc++.h>

using namespace std;

const int MAX = 2001;

int A[MAX],
    dp[MAX][MAX];

int solve(int s, int e)
{
    if (e <= s) return 1;
    if (dp[s][e] != -1) return dp[s][e];
    return dp[s][e] = A[s] == A[e] && solve(s+1, e-1);
}

int main()
{
    ios_base::sync_with_stdio(0);
    cin.tie(0);

    freopen("input.txt", "r", stdin);

    int N; cin >> N;
    for (int i = 1; i <= N; i++) cin >> A[i];
    int M; cin >> M;

    memset(dp, -1, sizeof dp);
    while (M--) {
        int s, e; cin >> s >> e;
        cout << solve(s, e) << '\n';
    }

    return 0;
}
